### PR TITLE
Add PR and issue templates

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description
Fix https://github.com/open-vsx/publish-extensions/issues/474

## Progress

- [x] Add issue templates
- [x] Add external issue template for requests
- [x] Add a PR template
- [x] Add a guide for publishing via GitHub CI
- [x] Add docs about problems with extensions (publishing/bugs within exts)
- [ ] Before merging ⚠️: change all the links to reflect the master (HEAD) branch

### CODE REVIEW

File: docs/external_contribution_request.md
Comment (loujaybee) [inline docs/external_contribution_request.md:11]: This PR: https://github.com/open-vsx/publish-extensions/pull/483/files and this PR are closely related. 

I think we should reference the other PR docs from within this issue template (we can do in a subsequent PR, we don't need to couple them together). But it's so that we make it super clear in these docs to the extension maintainers _why_ it's important that they should publish the extensions themselves.

Comment (filiptronicek) [reply docs/external_contribution_request.md:13]: Added a mention in [`87eee1b` (#479)](https://github.com/open-vsx/publish-extensions/pull/479/commits/87eee1b075673cc149db2ea33e0def14957827a4).

**Review by Siddhant-K-code — APPROVED** [review]
LGTM 🎉

File: docs/external_contribution_request.md
Comment (akosyakov) [inline docs/external_contribution_request.md:13]:
```
The docs for publishing the extension are [here](https://github.com/eclipse/openvsx/wiki/Publishing-Extensions). Technically this should be straightforward. You get a token and call `ovsx publish` with it. For example, this is how RedHat publishes their java extension: [redhat-developer/vscode-java@ff51dbf/Jenkinsfile#L73-L82](https://github.com/redhat-developer/vscode-java/blob/5d4f8d58b8e919534800ca7bc41e8513c288f573/Jenkinsfile#L78-L82)

[Here](https://github.com/PeterWone/vsc-print/pull/121) is also an example PR to contribute release GH action which can publish to GitHub, MS and OpenVSX at the same time.
```

Comment (akosyakov) [reply docs/external_contribution_request.md:13]: It is important to use permalinks.

File: docs/external_contribution_request.md
Comment (akosyakov) [inline docs/external_contribution_request.md:7]: Unfortunately MS prohibits usages of MS marketplace by any other products or redistribution vsix files from it. Because of it we kindly ask you to take the ownership of your namespace in [OpenVSX](https://open-vsx.org/) and publish there as well.

OpenVSX is a vendor neutral alternative to MS marketplace used by all other derivatives of VS Code like [VSCodium](https://vscodium.com/), [Gitpod](gitpod.io), [OpenVSCode](https://github.com/gitpod-io/openvscode-server), [Theia](https://theia-ide.org/) based IDEs, and so on.

The docs for publishing the extension are [here](https://github.com/eclipse/openvsx/wiki/Publishing-Extensions). Technically this should be straightforward. You get a token and call `ovsx publish` with it. For example, this is how RedHat publishes their java extension: [redhat-developer/vscode-java@ff51dbf/Jenkinsfile#L73-L82](https://github.com/redhat-developer/vscode-java/blob/5d4f8d58b8e919534800ca7bc41e8513c288f573/Jenkinsfile#L78-L82)

[Here](https://github.com/PeterWone/vsc-print/pull/121) is also an example PR to contribute release GH action which can publish to GitHub, MS and OpenVSX at the same time. Let @filiptronicek @akosyakov know if you are opened to a PR contributing such GH action. They would be glad to help.

Gitpod as well support open source developers by providing unlimited access to platform. Feel free to reach out via https://www.gitpod.io/docs/professional-open-source

Comment (akosyakov) [reply docs/external_contribution_request.md:7]: Let's try such template instead, please use permalinks as in text.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (akosyakov) [inline .github/PULL_REQUEST_TEMPLATE.md:17]: We should point here to template which they should use.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (akosyakov) [inline .github/PULL_REQUEST_TEMPLATE.md:8]: I don't think it is important. I would like to bring attention both on issues and PR for new or broken extensions to the fact that extension should be removed from this repo and published directly to OpenVSX by the author.

File: .github/ISSUE_TEMPLATES/bug.yml
Comment (akosyakov) [inline .github/ISSUE_TEMPLATES/bug.yml:1]: I think issue template should also tell go to the extension author and let's remove it from here. The bug itself means that we are not capable to maintain it well.

**Comment by akosyakov** [comment]
@filiptronicek Could you have a look at remaining comments? 🙏

File: docs/external_contribution_request.md
Comment (filiptronicek) [inline docs/external_contribution_request.md:15]: We could make this just the raw markdown (just a md code block) for easier copying.

**Comment by akosyakov** [comment]
@filiptronicek lgtm, don't forget to clean up the history before merging

Should not we also have a template for add new extension? Which again urge don't add it here? I think otherwise users will use another template which does not bring this problem?

**Comment by filiptronicek** [comment]
I have added the issue template, anything I should add, @akosyakov?

image

File: docs/external_contribution_request.md
Comment (loujaybee) [inline docs/external_contribution_request.md:7]: Any reason for wrapping in a code block here? As you don't get the markdown formatting?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (loujaybee) [inline .github/PULL_REQUEST_TEMPLATE.md:15]: I'd suggest we directly copy/paste the reasoning from the README for consistency, but also as it's more accurate, and also explains a bit more about _why_ it's better for authors to publish (and not just that it's a pain for us... which is true, that also!). But I'd lead with the benefits to the extension authors and/or the problems which occur when authors don't publish themselves. 

```suggestion
`publish-extensions` exists to seed the Open VSX marketplace, and also serves as a place for extensions that cannot feasibly be published directly by the extensions authors. In the long-run it is better for extension owners to publish their own plugins because:

1. Any future issues (features/bugs) with any published extensions in Open VSX will be directed to their original repo/source-control, and not confused with this repo publish-extensions.
2. Extensions published by official authors are shown within the Open VSX marketplace as such. Whereas extensions published via publish-extensions display a warning that the publisher (this repository) is not the official author.
3. Extension owners who publish their own extensions get greater flexibility on the publishing/release process, therefore ensure more accuracy/stability. For instance, in some cases publish-extensions has build steps within this repository, which can cause some uploaded plugin versions to break (e.g. if a plugin build step changes).

If you are the author of the extension being raised in this PR, please first consider directly publishing the extension yourself. You can refer to our [direct publish setup](docs/direct_publish_setup.md) doc for a guide on how to publish your plugin to Open VSX.
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (loujaybee) [inline .github/PULL_REQUEST_TEMPLATE.md:11]: I'd suggest we just remove this to take out some visual noise. 

```suggestion
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (loujaybee) [inline .github/PULL_REQUEST_TEMPLATE.md:2]: Could we move this point below some of the others, as it's not the main point/message we're trying to send?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (loujaybee) [inline .github/PULL_REQUEST_TEMPLATE.md:7]: Could we remove these comments? 

```suggestion
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (loujaybee) [inline .github/PULL_REQUEST_TEMPLATE.md:28]: Could we bring together this line and the checkbox? 

```suggestion
- [ ] I have tried reaching out to the extension maintainers about publishing this extension to OpenVSX (if no,  please create an issue in the extension's repo using [this template](docs/external_contribution_request.md).
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (loujaybee) [inline .github/PULL_REQUEST_TEMPLATE.md:34]:
```suggestion
This PR introduces X change for Y reason. 
```

File: docs/external_contribution_request.md
Comment (filiptronicek) [reply docs/external_contribution_request.md:7]: @loujaybee we decided it would be better for peeps who just go here to copy the template, so they don't copy just the rendered text.

File: docs/direct_publish_setup.md
Comment (loujaybee) [inline docs/direct_publish_setup.md:3]:
```suggestion
We advise extension authors to publish their extensions to OpenVSX as a part of their CI/CD process. See [our reasoning why that is](https://github.com/open-vsx/publish-extensions#when-to-add-an-extension).
```

File: docs/direct_publish_setup.md
Comment (loujaybee) [inline docs/direct_publish_setup.md:5]:
```suggestion
To make the process easier, we have provided a simple template of a GitHub Actions workflow to speed up the process of creating your own CI script/process. This template performs the following:
```

File: docs/direct_publish_setup.md
Comment (loujaybee) [inline docs/direct_publish_setup.md:9]:
```suggestion
- Manual CI triggers for Open VSX and/or Microsoft Marketplace publishing.
```

File: docs/direct_publish_setup.md
Comment (loujaybee) [inline docs/direct_publish_setup.md:14]:
```suggestion
1. The first step follows the [Publishing Extensions](https://github.com/eclipse/openvsx/wiki/Publishing-Extensions) doc and take note of the access token that you get as you will need it in the next step.
```

File: docs/direct_publish_setup.md
Comment (loujaybee) [inline docs/direct_publish_setup.md:16]:
```suggestion
2. To run the GitHub Action above, you need to setup two repository secrets for the Action to use:
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:5]:
```suggestion
If an extension you downloaded from OpenVSX isn't functioning as expected, please follow these steps to debug:
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:10]:
```suggestion
4. If they are the same version, visit the extension's source repository to investigate the issue further.
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:12]:
```suggestion
**Important:** We (`publish-extensions` maintainers) cannot help with functionally broken extensions; these issues should be raised with the respective maintainers.
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:14]:
```suggestion
If the extension has been unmaintained for a longer period of time, please [raise an issue](https://github.com/open-vsx/publish-extensions/issues/new) by selecting <kbd>Other</kbd> as the issue template and adding the necessary information (we will publish the last version by building manually and remove the extension from this repository).
```

File: docs/direct_publish_setup.md
Comment (filiptronicek) [reply docs/direct_publish_setup.md:9]: @loujaybee I think this new wording does not mention GitHub Releases, which are also a possibility. I am not sure how to rephrase it to still have the `and/or` part in it with three possible options 😄

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:18]:
```suggestion
If an extension on OpenVSX is outdated or has never been successfully published (despite being listed in [`extensions.json`](https://github.com/open-vsx/publish-extensions/blob/master/extensions.json)), the cause is likely one of the following:
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:27]:
```suggestion
The best way to solve issues of failed publishing is to publish the extension from its own repository, not this one. 
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:28]:
```suggestion
- If you are the extension author, you can use [this document](direct_publish_setup.md) as a guide on how to set up a CI publishing process.
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:29]:
```suggestion
- If you are a community member you can raise an issue (please check for existing issues) using [this template](external_contribution_request.md). If the maintainers are willing to accept a contribution you can use the [same document listed in the point above](direct_publish_setup.md) to quickly setup a CI job with GitHub Actions.
```

File: docs/external_contribution_request.md
Comment (loujaybee) [inline docs/external_contribution_request.md:16]: I'm not so sure on the Gitpod plug in this template, as we don't want Gitpod to be mistaken as the direct owner of Open VSX. I also wouldn't necessarily mention Anton/yourself. 

```suggestion
Dear extension author,

Please publish this extension to the Open VSX marketplace. 

**Context**
Unfortunately, as Microsoft prohibits usages of Microsoft marketplace by any other products or redistribution vsix files from it, in order to use VS Code extensions in non-Microsoft products, we kindly ask that you take ownership of the VS Code extension namespace in [OpenVSX](https://open-vsx.org/) and publish this extension on Open VSX.

**What is OpenVSX? Why does it exist?**
OpenVSX is a vendor neutral alternative to MS marketplace used by all other derivatives of VS Code like [VSCodium](https://vscodium.com/), [Gitpod](gitpod.io), [OpenVSCode](https://github.com/gitpod-io/openvscode-server), [Theia](https://theia-ide.org/) based IDEs, and so on.

**How can you publish to Open VSX?**
The docs to publish an extension can be found [here](https://github.com/eclipse/openvsx/wiki/Publishing-Extensions). This process is straightforward and shouldn't take too long. Essentially, you need an authentication token and to execute the command `ovsx publish` to publish your plugin. See this example from RedHat publishes their Java extension: [redhat-developer/vscode-java@ff51dbf/Jenkinsfile#L73-L82](https://github.com/redhat-developer/vscode-java/blob/5d4f8d58b8e919534800ca7bc41e8513c288f573/Jenkinsfile#L78-L82)

You can also find [an example PR](https://github.com/PeterWone/vsc-print/pull/121) that contributes a release GitHub Action which publishes to GitHub, MS and OpenVSX at the same time. 
```

File: .github/ISSUE_TEMPLATES/bug.yml
Comment (loujaybee) [inline .github/ISSUE_TEMPLATES/bug.yml:2]:
```suggestion
description: A bug report about a failure to publish via the open-vsx/publish-extensions repository
```

File: .github/ISSUE_TEMPLATES/bug.yml
Comment (loujaybee) [inline .github/ISSUE_TEMPLATES/bug.yml:17]:
```suggestion
  - type: textarea
    id: describe
    attributes:
      label: Expected result and actual result
      description: What actually occurred, and what did you expect?
      value: "I expected X plugin to publish successfully, but it is currently intermittently failing"
    validations:
      required: true
```

File: .github/ISSUE_TEMPLATES/extension.yml
Comment (loujaybee) [inline .github/ISSUE_TEMPLATES/extension.yml:11]:
```suggestion
      description: We require new publication requests to be first communicated with the extension's maintainers (see [why](https://github.com/open-vsx/publish-extensions/blob/master/README.md#when-to-add-an-extension)), it the author cannot publish the extension or is unresponsive, raise a PR. If you haven't made an issue for the extension owner yet, please create one in the extension's repository using [this template](https://github.com/open-vsx/publish-extensions/blob/master/docs/external_contribution_request.md).
```

File: docs/direct_publish_setup.md
Comment (loujaybee) [reply docs/direct_publish_setup.md:9]: Updated! Does it work now? 😁

File: docs/direct_publish_setup.md
Comment (loujaybee) [inline docs/direct_publish_setup.md:12]:
```suggestion
## Setup VS Code extension publishing CI workflow
```

File: docs/extension_issues.md
Comment (loujaybee) [inline docs/extension_issues.md:3]:
```suggestion
## Functional bugs in VS Code extensions
```

File: docs/external_contribution_request.md
Comment (loujaybee) [inline docs/external_contribution_request.md:3]: Should we just use this as the title, otherwise the copy/paster has to update/remove the placeholder title? 

```suggestion
## Publish `EXTENSION_NAME` to Open VSX

```

File: docs/external_contribution_request.md
Comment (loujaybee) [reply docs/external_contribution_request.md:7]: I'd then remove all the boilerplate and simply keep the file to copy/paste the full thing? 🤔  

WDYT?

File: docs/direct_publish_setup.md
Comment (filiptronicek) [reply docs/direct_publish_setup.md:9]: I think this will work, although it does not mention GitHub Releases explicitly, but it is stated two points above, so it isn't all that necessary.

File: docs/external_contribution_request.md
Comment (filiptronicek) [reply docs/external_contribution_request.md:16]: This looks like a great change, @loujaybee! I would go so far as to add a section linking to the https://github.com/open-vsx/publish-extensions/blob/ft/issue-and-pr-templates/docs/direct_publish_setup.md doc and add a mention of us (or the person that creates the PR) being open to contributing the CI workflow to the target repository.

The idea with the mentions was by @akosyakov, so he probably has some thoughts 😃.

File: docs/external_contribution_request.md
Comment (filiptronicek) [reply docs/external_contribution_request.md:3]: Hmm I think it wouldn't exactly make sense this way, because there is also a heading saying `Issue Body`. I think it will be even better if we make the issue title a code block the same way as the issue body.

We could also simply have the whole md file be the issue content (make the first heading imply to be used as the issue title), but in this way we can both add notes to the sections (maybe some guidance to the contributors), as well as providing a simple copying experience thanks to the code blocks.

File: docs/external_contribution_request.md
Comment (akosyakov) [reply docs/external_contribution_request.md:16]: I'm fine to drop it. Put Gitpod there to motivate author a bit to become OpenVSX users through it.

Comment (brianking) [reply docs/external_contribution_request.md:16]: > I'm not so sure on the Gitpod plug in this template, as we don't want Gitpod to be mistaken as the direct owner of Open VSX. I also wouldn't necessarily mention Anton/yourself.

I agree. As a possible alternative, I am happy to collaborate on a resource that points people to ways they can try Open VSX. One idea is a page on open-vsx.org, a la https://github.com/EclipseFdn/open-vsx.org/issues/773

File: docs/external_contribution_request.md
Comment (filiptronicek) [reply docs/external_contribution_request.md]: I implemented it as I said here, but we can change it anytime if it does not make sense to you 🙂.

**Comment by akosyakov** [comment]
good to merge?

**Comment by filiptronicek** [comment]
I think so!

File: docs/exampleCI.yaml
Comment (Siddhant-K-code) [inline docs/exampleCI.yaml:117]: we can add a newline at EOF

File: .github/ISSUE_TEMPLATES/config.yml
Comment (Siddhant-K-code) [inline .github/ISSUE_TEMPLATES/config.yml:5]: we can add a newline at EOF

**Review by Siddhant-K-code — COMMENTED** [review]
adding a newline at EOF

File: docs/exampleCI.yaml
Comment (filiptronicek) [reply docs/exampleCI.yaml:117]: Thank you!

**Pull request merged**
